### ISSUE TITLE
Idea: make `Pretrained` a subclass of nn.Module?

### ISSUE THREAD

**Issue by PiotrDabkowski** [speechbrain/speechbrain#987]
Currently the `Pretrained` class is a parent of all the pretrained modules, however inference is a bit tricky and non-standard as it does not inherit from `torch.nn.Module`.

For example, to move the model to a selected device you would need to specify run_opts during instantiation. Same for the eval mode. Finally, reusing these modules as part of a larger PyTorch model is again not trivial. It is possible to work around all these difficulties, but why?

Simply inheriting from torch.nn.Module would solve these inconveniences:
`Pretrained (torch.nn.Module):`

is there a specific reason why this was not done?

**Comment by TParcollet** [comment]
Hi that is a good question for our very own @Gastron.

**Comment by Gastron** [comment]
Good idea, I see how inheriting from torch.nn.Module could make `Pretrained` classes a little easier to use as part of a larger system. I don't remember if we had any insurmountable reason not to do that in the first place.

There is at least one issue: what to put in the `forward` method of each `Pretrained` subclass interface, since basically all subclasses have multiple methods that do something of interest. I think for all of these we can find a good choice, though. Basically `forward` should work on batch-tensors and apply the model as completely as possible (e.g. encode-decode instead of simply encode).

**Comment by TParcollet** [comment]
I agree, we would also need to do some refactoring on the HuggingFace side. @PiotrDabkowski would you please provide an example with one of the interface? We don't have the resources to put someone on that right now, but we should consider it.

**Comment by Gastron** [comment]
What I like here is that we should be able to make this change without breaking anything existing. I wonder if there could be future `Pretrained` applications where a `torch.nn.Module` interface is not suitable?

**Comment by TParcollet** [comment]
I don't think so, maybe we should set this up for the next SB version. @Gastron what do you think ?

**Comment by Gastron** [comment]
Let's try it out, sure
